helm/hpcc/templates/_helpers.tpl

```diff
@@ -2058,3 +2058,31 @@ spec:
 {{- end -}}
 {{- end -}}
 
+{{- define "hpcc.generateWarnings" -}}
+{{/*
+Generate a list of warnings
+Pass in dict with root
```

Comment:
Not clear what root is for.  To allow extra items to be passed in?  Otherwise I don't think it is used (or I don't understand) - .root is not used within the function.

Reply:
It is passed down to getWarnings() and it is needed to generate the warnings.

Reply:
Why?  $ctx is passed, and warnings is filled in.  But what does merging .root do?  trying follow the logic.

Reply:
_warnings.tpl (line 31) includes helper function hpcc.imageName.  And hpcc.imageName uses .root.Values.global.image.name.  If I don't pass root, I get the following error: 

Error: template: ...error calling include: template: hpcc/templates/_helpers.tpl:1825:25: executing "hpcc.filterConfig" at <include .configMapHelper .>: error calling include: template: hpcc/templates/esp.yaml:79:3: executing "hpcc.espConfigMap" at <include "hpcc.generateWarnings" .>: error calling include: template: hpcc/templates/_helpers.tpl:2067:4: executing "hpcc.generateWarnings" at <include "hpcc.getWarnings" $ctx>: error calling include: template: hpcc/templates/_warnings.tpl:31:22: executing "hpcc.getWarnings" at <include "hpcc.imageName" (dict "root" $ "me" dict)>: error calling include: template: hpcc/templates/_helpers.tpl:698:8: executing "hpcc.imageName" at <.root.Values.global.image.root>: nil pointer evaluating interface {}.global

Reply:
Agree, it should pass in the root dict as "root":
```
{{- $ctx := dict "warnings" list "root" .root -}}
```

and then reference "root" in getWarnings.
But that means changing all lines that reference . to .root. e.g.:
```
 {{- $defaultVersion := .Values.global.image.version | default .Chart.Version -}}
```
should change to:
```
 {{- $defaultVersion := .Values.global.image.version | default .root.Chart.Version -}}
```

Also, the last line where it's setting "warnings" into the returned ctx, that should be changed to:
```
 {{- $_ := set . "warnings" $warnings -}}
```

since the passed in ctx is . 

and remove the:
```
 {{- $root := . -}}
```

Reply:
This still isn't right.  Now you are passing in . as the parameter, so there is no .root.

I think you need to
i) go back to passing a dictionary as a parmeter
ii) use $ctx := dict "warnings" list "root" .root
iii) delete the definition of $defaultImage in the warning code
iv) Use .root.x where you previously have .x e.g. .root.Chart.version, .root.Values.storage

I still don't quite understand what $ means @jakesmith can you clarify when that changes?

Reply:
$ changes within an include template function to the dictionary passed to it.
So by convention we pass in the root context ($) as "root" and use it within our template functions